SofaKernel/modules/SofaHelper/src/sofa/helper/NameDecoder.h

```diff
@@ -0,0 +1,263 @@
+ *      - damien.marchal@univ-lille.fr
+ *      - jean-nicolas.brunet@inria.fr
+ *****************************************************************************/
+#pragma once
+
+#include <string>
+#include <typeinfo>
+#include <tuple>
+#include <sofa/helper/helper.h>
+namespace sofa::helper
+{
+
+template<class T>
+class HasGetCustomTemplateName
+{
+    typedef char YesType[1];
+    typedef char NoType[2];
+
+    template<typename C> static YesType& test( decltype (&C::GetCustomTemplateName) );
+    template<typename C> static NoType& test(...);
+
+public:
+    enum { value = sizeof(test<T>(0)) == sizeof(YesType) };
+};
+
+template<class T>
+class HasNameFunction
+{
+    typedef char YesType[1];
+    typedef char NoType[2];
+
+    template<typename C> static YesType& test( decltype (&C::Name) );
+    template<typename C> static NoType& test(...);
+
+public:
+    enum { value = sizeof(test<T>(0)) == sizeof(YesType) };
+};
+
+
+template<class T>
+class HasGetCustomClassName
+{
+    typedef char YesType[1];
+    typedef char NoType[2];
+
+    template<typename C> static YesType& test( decltype (&C::GetCustomClassName) );
+    template<typename C> static NoType& test(...);
+
+public:
+    enum { value = sizeof(test<T>(0)) == sizeof(YesType) };
+};
+
+//TODO(dmarchal: 01/04/2020) This is for compatibility layer, remove it after 01/01/2021
+template<class T>
+class HasGetDefaultTemplateName
+{
+    typedef char YesType[1];
+    typedef char NoType[2];
+
+    template<typename C> static YesType& test( decltype (&C::GetDefaultTemplateName) );
+    template<typename C> static NoType& test(...);
+
+public:
+    enum { value = sizeof(test<T>(0)) == sizeof(YesType) };
+};
+
+//TODO(dmarchal: 01/04/2020) This is for compatibility layer, remove it after 01/01/2021
+template<class T>
+class HasDeprecatedTemplateName
+{
+    typedef char YesType[1];
+    typedef char NoType[2];
+
+    template<typename C> static YesType& test( decltype (&C::templateName) );
+    template<typename C> static NoType& test(...);
+
+public:
+    enum { value = sizeof(test<T>(0)) == sizeof(YesType) };
+};
+
+//TODO(dmarchal: 01/04/2020) This is for compatibility layer, remove it after 01/01/2021
+template<class T>
+class HasDeprecatedClassName
+{
+    typedef char YesType[1];
+    typedef char NoType[2];
+
+    template<typename C> static YesType& test( decltype (&C::className) );
+    template<typename C> static NoType& test(...);
+
+public:
+    enum { value = sizeof(test<T>(0)) == sizeof(YesType) };
+};
+
+template<class T>
+class HasName
+{
+    typedef char YesType[1];
+    typedef char NoType[2];
+
+    template<typename C> static YesType& test( decltype (&C::Name) );
+    template<typename C> static NoType& test(...);
+
+public:
+    enum { value = sizeof(test<T>(0)) == sizeof(YesType) };
+};
+
+template<typename T>
+std::string GetSofaTypeTemplateName(const std::string prefix="");
+
+template<typename T, typename ...Ts>
+std::enable_if_t<sizeof...(Ts)!=0, std::string>
```

Comment:
Small readability improvement, I think we should avoid `enable_if` where it's possible:
```cpp
template<typename T1, typename T2, typename ...Ts>
std::string GetSofaTypeTemplateName(const std::string prefix="") {
    return GetSofaTypeTemplateName<T1>(prefix) + GetSofaTypeTemplateName<T2, Ts...>(",");
} 
```